### ISSUE TITLE
Visual Studio 2015 build succeeds even with TypeScript error when generated JS is present.

### ISSUE THREAD

**Issue by onlyann** [Microsoft/TypeScript#7456]
I am using Visual Studio 2015 Update 1 and Typescript for VS2015 1.8.4.
I have a project that contains TypeScript files.
Here's my `tsconfig.json`
```
{
  "compileOnSave": true,
  "compilerOptions": {
    "target": "es5",
    "module": "system",
    "moduleResolution": "node",
    "sourceMap": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": false,
    "noImplicitAny": false,
    "noEmitOnError": true
  },
  "exclude": [
    "node_modules",
    "typings"
  ]
}
```

When I write incorrect TypeScript, let's say i have `file1.ts` and write `foo` , I obtain the error message
```Cannot find name 'foo'```

If I try to build the project, it will result in a build failure. When the `file1.ts` file is correct, it generates the `file.js` file next to it and I can build successfully. So far so good.

Now, if I change once again `file1.ts` , it will show `Cannot find name 'foo'` in the error list but building the project succeeds.

What I expect is for the project build to fail.

The project build will fail again when I remove the generated `file1.js` file.

Shouldn't the TypeScript compiler try to recompile all the *.ts files that are newer than their output when I perform a build, or have the `compileOnSave` option to remove the output?

**Comment by billti** [comment]
The current design is that we still emit in the presence of type errors. Only syntax errors (i.e. invalid JavaScript - regardless of type information) cause the file not to be emitted. Note the build should still return the error, but it should also emit runnable code.

**Comment by onlyann** [comment]
@billti I don't follow your explanations. I am describing the introduction of a syntax error in the TypeScript file and that the MSBuild target doesn't fail in the presence of an existing generated JavaScript file with a past timestamp.

**Comment by billti** [comment]
I'm still finding your description a little hard to follow, but if I understand correctly, then I may have a hunch as to what is going on.

When we compile-on-save, if the output is going to be identical to what already exists, then we don't re-emit in order to save time (writing files is expensive). In this case perhaps, depending on the changes you made, if the output would have been the same as the already existing file on disk (i.e. you only introduced type errors), then if we don't re-emit that file we also don't generate the errors for that file.

Does this sound like a possible explanation for the scenario you attempted?

+ @mhegazy and @vladima who are more familiar with these optimizations.

**Comment by mhegazy** [comment]
this is a duplicate of https://github.com/Microsoft/TypeScript/issues/7247. When using tsconfig.json, the MSBuild targets do not know the outputs, currently it means that the project is up-to-date. which is obviously wrong in this case. 
Compile-on-Save should work correctly, in the case above "noEmitOnError" is set to true, so saving does not regenerate the file.
rebuild should always work as well.

**Comment by onlyann** [comment]
@mhegazy you are spot on. This is precisely #7247 where it is explained much better. Sorry for the duplicate, I didn't find that issue before opening this one.
